src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -231,28 +233,27 @@ public void StartTagHelperWritingScope()
         /// <returns>The buffered <see cref="TagHelperContent"/>.</returns>
         public TagHelperContent EndTagHelperWritingScope()
         {
-            if (_writerScopes.Count == 0)
+            if (_tagHelperScopes.Count == 0)
             {
                 throw new InvalidOperationException(Resources.RazorPage_ThereIsNoActiveWritingScopeToEnd);
             }
 
-            var writer = _writerScopes.Pop();
-            Debug.Assert(writer == ViewContext.Writer);
+            // Reviewers: Debug.Assert meant we weren't supporting user changes to ViewContext.Writer in any case.
+            // Asserting about a user behavior also just doesn't make sense. I chose to silently ignore that case
+            // instead. Other reasonable options are to (a) throw if ViewContext.Writer is not of expected type or
+            // (b) go back to pushing changed values in Start..., also throw above when Count is 1, Pop and throw if
+            // values are unexpected, and Peek() normally but Pop() in the restoration code below when Count was 2.
 
-            if (_writerScopes.Count > 0)
-            {
-                ViewContext.Writer = _writerScopes.Peek();
-            }
-            else
-            {
-                ViewContext.Writer = _originalWriter;
+            // Get the content written during the current scope.
+            var writer = ViewContext.Writer as HtmlContentWrapperTextWriter;
+            var tagHelperContent = new DefaultTagHelperContent();
+            tagHelperContent.AppendHtml(writer?.ContentBuilder);
```

Comment:
If `writer` is `null` you'll have a `null` entry in `tagHelperContent`. Would be better to just not `AppendHtml` in that case.